### PR TITLE
Mapping for Hercules DJControl Inpulse 200

### PR DESCRIPTION
**Pull request by DJPhatso**
Initial release of MIDI configuration files for Hercules DJControl Inpulse 200

### CODE REVIEW

**Comment by uklotzde** [comment]
Ready for merge? Code is readable, Wiki page exists.

File: Hercules-DJControl-Inpulse-200-script.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+// * Mixxx mapping script file for the Hercules DJControl Inpulse 200.
+// * Author: DJ Phatso
+// * Version 1.0 (March 2019)
+// * Forum: https://www.mixxx.org/forums/viewtopic.php?f=7&t=12592
+// * Wiki: https://mixxx.org/wiki/doku.php/hercules_djcontrol_inpulse_200
+
+// TODO: Functions that could be implemented to the script:
+//
+// * FX:
+//  - See how to preselect efffects for a rack
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js]: effects

File: Hercules-DJControl-Inpulse-200-script.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+var DJCi200 = {};
+
+DJCi200.scratchButtonStateDA = true;
+DJCi200.scratchButtonStateDB = true;
+
+
+DJCi200.init = function() {
+
+
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js]: IMHO you went a bit overboard with all that newlines throughout the script, but that's just a preference.

File: Hercules-DJControl-Inpulse-200-script.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    var newValue;
+    if (value < 64) {
+        newValue = value;
+    } else {
+        newValue = value - 128;
+    }
+
+    {
+        engine.setValue('[Channel' + 1 + ']', 'jog', newValue); // Pitch bend
+    }
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js]: Why the curly brackets? Is there supposed to be some if condition that is missing?

File: Hercules-DJControl-Inpulse-200-script.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    var newValue;
+    if (value < 64) {
+        newValue = value;
+    } else {
+        newValue = value - 128;
+    }
+
+    {
+        engine.setValue('[Channel' + 2 + ']', 'jog', newValue); // Pitch bend
+    }
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js]: same (brackets)

File: Hercules-DJControl-Inpulse-200-script.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+
+
+// ****************************************************************************
+function DJCi200() {}
+
+var DJCi200 = {};
+
+DJCi200.scratchButtonStateDA = true;
+DJCi200.scratchButtonStateDB = true;
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js]: A more descriptive name like `scratchButtonEnabledDA` would be nice.

File: Hercules-DJControl-Inpulse-200-script.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        engine.scratchTick(2, newValue); // Scratch!
+    }
+};
+
+
+
+DJCi200.shutdown = function() {
+
+    midi.sendShortMsg(0xB0, 0x7F, 0x00);
+};
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js]: Please add a newline at end of file. Most editors do this automatically, which would pollute patch diffs later on. So better at it now ;-)

File: Hercules_DJControl_Inpulse_200.midi.xml [hunk 7]
```diff
@@ -0,0 +1,2020 @@
...
+				<midino>0x33</midino>
+				<on>0x7F</on>
+				<minimum>0.5</minimum>
+			</output>
+
+
+		</outputs>
+
+	</controller> 
+</MixxxMIDIPreset>
```
Comment (Holzhaus) [inline Hercules_DJControl_Inpulse_200.midi.xml]: Please add a newline at the end of the file.

File: Hercules_DJControl_Inpulse_200.midi.xml [hunk 8]
```diff
@@ -0,0 +1,2020 @@
...
+			<file filename="midi-components-0.0.js"/>
+			<file functionprefix="DJCi200" filename="Hercules-DJControl-Inpulse-200-script.js" />
+		</scriptfiles>
+		<controls>
+
+
+			/****************************************************************************************************/
+			// NN's MIDI Channel 1 (0x90)
+			/****************************************************************************************************/
+
```
Comment (Holzhaus) [inline Hercules_DJControl_Inpulse_200.midi.xml]: These aren't comments but XML text nodes. Mixxx' mapping XML parser might currently allow this, but please use actual XML comments here (`<!-- foobar -->`).

**Review by Holzhaus — CHANGES_REQUESTED** [review]

**Comment by Be-ing** [comment]
@Holzhaus thanks for jumping in here. @DJPhatso I apologize for putting off reviewing your pull requests (as well as some other controller mappings) for so long. I have some time now and will go through them. Feel free to leave a comment to remind us that you're waiting on a review if it's been a long time with no activity.

**Comment by DJPhatso** [comment]
Hi guys,

You might want to hold off reviewing for the moment. Following all the changes and code cleanup made by @kerrickstaley for the DJControl Starlight mapping (which has been approved), I've just started applying the same changes to the Inpulse 200 & 300 mappings.

**Comment by DJPhatso** [comment]
Mapping files for Hercules DJControl Inpulse 200

Changes to v1.2

Code cleanup.
Changes to v1.1

Fix seek-to-start and cue-master behavior.
Tweak scratch, seek, and bend behavior (User settigns).
Controller knob/slider values are queried on startup, so MIXXX is synced.
Fixed vinyl button behavior the first time it's pressed.



v1.0 : Original forum release

**Comment by Holzhaus** [comment]
Hi, can you rebase on 2.2? This could also be part of the 2.2.4 release.

File: Hercules-DJControl-Inpulse-200-script.js [hunk 9]
```diff
@@ -0,0 +1,249 @@
+////////////////////////////////////////////////////////////////////////
+// JSHint configuration                                               //
+////////////////////////////////////////////////////////////////////////
+/* global engine                                                      */
+/* global script                                                      */
+/* global print                                                       */
+/* global midi                                                        */
+////////////////////////////////////////////////////////////////////////
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js:8]: Please remove the JSHint config, this conflicts with eslint that we're using now.

File: Hercules-DJControl-Inpulse-200-script.js [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+// See this GitHub issue for more context:
+// https://github.com/eslint/eslint/issues/1939
+// *eslint-disable no-unused-vars*/
+
+DJCi200.kScratchActionNone = 0;
+DJCi200.kScratchActionScratch = 1;
+DJCi200.kScratchActionSeek = 2;
+DJCi200.kScratchActionBend = 3;
+
+//function DJCi200() {}
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js:68]: Please remove this line

File: Hercules-DJControl-Inpulse-200-script.js [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    engine.setParameter("[EffectRack1_EffectUnit1_Effect2]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit1_Effect3]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit2_Effect1]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit2_Effect2]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit2_Effect3]", "meta", 0.6);
+    engine.setParameter("[EffectRack1_EffectUnit1]", "mix", 1);
+    engine.setParameter("[EffectRack1_EffectUnit2]", "mix", 1);
+
+
+
```
Comment (Holzhaus) [inline Hercules-DJControl-Inpulse-200-script.js:101]: I think a single newline suffices here (and also in other parts of the mapping).

**Review by Holzhaus — CHANGES_REQUESTED** [review]

**Comment by DJPhatso** [comment]
I thought I had closed this one already. 
#2542 was started and completed and already been merged.

**Pull request closed without merging**